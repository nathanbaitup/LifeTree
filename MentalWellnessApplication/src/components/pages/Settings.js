import React, { useState, useEffect } from 'react';
import { Text, View, ImageBackground, Button, Image, ActivityIndicator, StyleSheet } from 'react-native';

import ImagePicker, { launchCamera, launchImageLibrary } from 'react-native-image-picker';

import storage from '@react-native-firebase/storage';

export default function Settings(props) {

    // Parsing the logout function from App.js
    const logout = props.logout;

    const userID = props.extraData.id;

    const [uploading, setUploading] = useState(false);
    const [image, setImage] = useState(null);
    const [transferred, setTransferred] = useState(0);

    // REFERENCE ACCESSED 16/12/2021 https://www.instamobile.io/mobile-development/react-native-firebase-storage/
    // Used to be able to use the camera and image library of the device to capture / select an image to use for the profile picture of the user and save that image to storage in firebase.

    const takeImage = () => {
        // Sets how the image should be saved to firebase and the path to save it at.
        const options = {
            maxWidth: 2000,
            maxHeight: 2000,
            storageOptions: {
                skipBackup: true,
                path: '/images/userProfilePictures/'
            }
        };

        // Launch the camera for the device and if the image was taken, then set the image to the image uri.
        launchCamera(options, response => {
            if (response.didCancel) {
                console.log('User cancelled image picker');
            } else if (response.error) {
                console.log('ImagePicker Error: ', response.error);
            } else if (response.customButton) {
                console.log('User tapped custom button: ', response.customButton);
            } else {
                const source = { uri: response.assets[0].uri };
                setImage(source);
            }
        });
    }

    const selectImage = () => {
        // Sets how the image should be saved to firebase and the path to save it at.
        const options = {
            maxWidth: 2000,
            maxHeight: 2000,
            storageOptions: {
                skipBackup: true,
                path: '/images/userProfilePictures/'
            }
        };

        // Launch the image library for the device and if the image was selected, then set the image to the image uri.
        launchImageLibrary(options, response => {
            if (response.didCancel) {
                console.log('User cancelled image picker');
            } else if (response.error) {
                console.log('ImagePicker Error: ', response.error);
            } else if (response.customButton) {
                console.log('User tapped custom button: ', response.customButton);
            } else {
                const source = { uri: response.assets[0].uri };
                setImage(source);
            }
        });
    }

    // The uploadImage function, that sets the image uri and filename, then uploads the file to firebase storage,
    // and sets the image state back to null.
    const uploadImage = async () => {
        const { uri } = image;
        const filename = userID;
        const uploadUri = Platform.OS === 'ios' ? uri.replace('file://', '') : uri;

        setUploading(true);
        setTransferred(0);

        const task = storage()
        .ref(filename)
        .putFile(uploadUri);

        task.on('state_changed', snapshot => {
        setTransferred(
          Math.round(snapshot.bytesTransferred / snapshot.totalBytes) * 10000
        );
      });

        try {
            await task;
        } catch (error) {
            console.error(error);
        }
        setUploading(false);
        alert('Successfully uploaded image.');
        setImage(null);
    }

    return (
        <ImageBackground source={require('../resources/img/background.png')} style={{ width: '100%', height: '100%', opacity: 50 }} >
            <View style={styles.contentContainer}>
                <Text> Settings Page </Text>
                {image !== null ? (
                    <Image source={{ uri: image.uri }} style={styles.imageBox} />
                ) : null}
                <Button title='Take a profile picture' onPress={takeImage} />
                <Button title='Select a profile picture' onPress={selectImage} />

                <Button title='Upload Image' onPress={uploadImage} />
                <Button title='Logout' onPress={logout} />

            </View>
        </ImageBackground>
    );
}

// The styling for the settings page.
const styles = StyleSheet.create({
    contentContainer: {
        margin: 20,
        padding: 10,
        backgroundColor: '#FFFFFF',
        borderRadius: 25,
    },
    imageBox: {
        width: 300,
        height: 300,
        borderRadius: 150,
    },

});
