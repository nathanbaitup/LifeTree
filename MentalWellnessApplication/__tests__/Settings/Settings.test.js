import 'react-native';
import React from 'react';
import Settings from '../../src/components/pages/Settings/Settings';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('Checks to see if the Settings page renders', () => {
  const tree = renderer.create(<Settings />).toJSON();
  expect(tree).toMatchSnapshot();
});