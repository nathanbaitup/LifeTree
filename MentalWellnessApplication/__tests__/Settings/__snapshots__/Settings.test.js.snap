// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checks to see if the Settings page renders 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": "100%",
      "opacity": 50,
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/resources/img/background.png",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 25,
        "height": 988.148148148148,
        "margin": 20,
        "padding": 10,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "flex-start",
          "opacity": 1,
          "paddingBottom": 10,
          "paddingTop": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontSize": 16,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            Object {
              "color": "#448aff",
            },
          ]
        }
      >
         
        &gt;
         Return Home
      </Text>
    </View>
    <RCTScrollView>
      <View>
        <Text
          style={
            Object {
              "color": "#000000",
              "fontSize": 16,
              "fontWeight": "bold",
              "paddingTop": 15,
            }
          }
        >
          Profile Picture:
        </Text>
        <Text
          style={
            Object {
              "color": "#000000",
              "fontSize": 14,
              "paddingBottom": 10,
            }
          }
        >
          Select a picture from your library or take a picture:
        </Text>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#00e676",
              "borderRadius": 15,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginTop": 5,
              "opacity": 1,
              "padding": 10,
              "width": 250,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000000",
                "fontSize": 16,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Take a picture
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#00e676",
              "borderRadius": 15,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginTop": 5,
              "opacity": 1,
              "padding": 10,
              "width": 250,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000000",
                "fontSize": 16,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Select a profile picture
          </Text>
        </View>
        <View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#00e676",
                "borderRadius": 15,
                "justifyContent": "center",
                "marginBottom": 5,
                "marginTop": 5,
                "opacity": 1,
                "padding": 10,
                "width": 250,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "textAlign": "center",
                }
              }
            >
              Upload Image
            </Text>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "#000000",
              "fontSize": 16,
              "fontWeight": "bold",
              "paddingTop": 15,
            }
          }
        >
          Other Settings:
        </Text>
        <View
          accessible={true}
          collapsable={false}
          focusable={false}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#f57367",
              "borderRadius": 15,
              "justifyContent": "center",
              "marginTop": 10,
              "opacity": 1,
              "padding": 15,
              "width": 250,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000000",
                "fontSize": 16,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Logout
          </Text>
        </View>
        <View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#bababa",
                  "paddingTop": 30,
                  "textAlign": "center",
                }
              }
            >
              Inspirational quotes provided by: 
            </Text>
            <Text
              onPress={[Function]}
              style={
                Object {
                  "color": "blue",
                  "textAlign": "center",
                  "textDecorationLine": "underline",
                }
              }
            >
              ZenQuotes API
            </Text>
          </View>
          <Text
            style={
              Object {
                "color": "#bababa",
                "paddingTop": 30,
                "textAlign": "center",
              }
            }
          >
             LifeTree v1.1.0 2021
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
